#!/bin/bash

#SBATCH --job-name=SeqsXlocus     ## job name
#SBATCH -e SeqsXlocus_%j.e.txt    ## error message name
#SBATCH -o SeqsXlocus.log.%j.out  ## log file output name
#SBATCH -p standard
#SBATCH -c 1
#SBATCH --ntasks=1

#######################
#The script loops through each line of the text file and uses grep to find the matching headers in the fasta file. The -A 1 option tells grep to also print the next line (which will be the sequence) after each matched header. The output is then appended to the output file.

#### --- User-defined variables ---

INDIR=/home/pmorin/projects/Miscellaneous/TEST_phylogenomics_extract_loc_align_repo
LOCDIR=${INDIR}/loci_out
OUTDIR=${INDIR}/aligned_loci_out
mkdir -p ${OUTDIR}


BW_LOC=${LOCDIR}/"UCE_loc_list.txt" 
#This is just the "description" column from the bed file, in a text file.

#### --- end User-defined variables ---

dos2unix $BW_LOC # this is needed if the text file has returns that are being included in the array. This converts dos returns to unix returns so they aren't included.

ls -1 ${LOCDIR}/*bedout.fa > ${LOCDIR}/Ziph_bedout.fa_list.txt
# this is a list of fasta files containing the selected sequences from each species, generated by the script "BW_extract_genes_array_Sedna.sh"
BW_SEQ="Ziph_bedout.fa_list.txt" #FASTALIST="fastalist.txt"

# Read each line of the text file into an array
mapfile -t TEXTARRAY < $BW_LOC

# Loop through each fasta file in the list to copy sequences to output file for each locus.
while read -r FASTAFILE
do
    for i in "${!TEXTARRAY[@]}"
    do
    OUTFILE="Loc_$i.fasta"
    echo $OUTFILE
    echo $FASTAFILE
    echo "${TEXTARRAY[i]}"
        # Find the matching headers in the fasta file and copy matching headers and following line (sequence) to the output file
        grep -A 1 "${TEXTARRAY[i]}:" $FASTAFILE >> ${OUTDIR}/$OUTFILE
    done
done < $LOCDIR/$BW_SEQ 

